addons/pos_cache/models/pos_cache.py

```diff
@@ -0,0 +1,64 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from datetime import datetime
+import cPickle
+
+from openerp import models, fields, api
+
+class pos_cache(models.Model):
+    _name = 'pos.cache'
+
+    cache = fields.Binary()
+    cachetime = fields.Datetime()
+    config_ids = fields.One2many('pos.config', 'cache_id', 'Config')
+
+    @api.model
+    def refresh_all_caches(self):
+        caches = self.search([('config_ids', '=', False)])
+        if caches:
+            caches.unlink()
+        self.env['pos.config'].search([('state', '=', 'active')]).recomputeCache()
+
+    @api.one
+    def refresh_cache(self):
+        products = self.env['product.product'].search(self._get_domain())
+        prod_ctx = products.with_context(pricelist=self.config_ids[0].pricelist_id.id, display_default_code=False)
+        if self.config_ids[0].cachecomputeuser_id:
+            prod_ctx = prod_ctx.sudo(self.config_ids[0].cachecomputeuser_id.id)
+        res = prod_ctx.read(self._get_fields())
+        datas = {
+            'cache': cPickle.dumps(res, protocol=cPickle.HIGHEST_PROTOCOL),
+            'cachetime': fields.Datetime.now(),
+        }
+
+        self.write(datas)
+
+    def _get_domain(self):
+        return [('sale_ok', '=', True), ('available_in_pos', '=', True)]
+
+    def _get_fields(self):
+        return ['display_name', 'list_price','price','pos_categ_id', 'taxes_id', 'barcode', 'default_code',
+                'to_weight', 'uom_id', 'description_sale', 'description',
+                'product_tmpl_id']
+
+
+class pos_config(models.Model):
+    _inherit = 'pos.config'
+
+    #Use a related model to avoid the load of the cache when the pos load his config
+    cache_id = fields.Many2one('pos.cache', 'Cache')
+    cachecomputeuser_id = fields.Many2one('res.users', 'Cache compute user', help="User used to compute the cache content and thus avoid the bad practice of using admin in multi-company")
+    cachetime = fields.Datetime(related='cache_id.cachetime', string='Last cache compute')
+
+    @api.multi
+    def getProductsFromCache(self):
```

Comment:
Done

On 22/09/15 13:34, antonylesuisse wrote:
>
> In addons/pos_cache/models/pos_cache.py 
> <https://github.com/odoo/odoo/pull/8652#discussion_r40075748>:
>
> > +    def _get_fields(self):
> > +        return ['display_name', 'list_price','price','pos_categ_id', 'taxes_id', 'barcode', 'default_code',
> > +                'to_weight', 'uom_id', 'description_sale', 'description',
> > +                'product_tmpl_id']
> > +
> > +
> > +class pos_config(models.Model):
> > +    _inherit = 'pos.config'
> > +
> > +    #Use a related model to avoid the load of the cache when the pos load his config
> > +    cache_id = fields.Many2one('pos.cache', 'Cache')
> > +    cachecomputeuser_id = fields.Many2one('res.users', 'Cache compute user', help="User used to compute the cache content and thus avoid the bad practice of using admin in multi-company")
> > +    cachetime = fields.Datetime(related='cache_id.cachetime', string='Last cache compute')
> > +
> > +    @api.multi
> > +    def getProductsFromCache(self):
>
> use _ instead of camel convention for python methods
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/odoo/odoo/pull/8652/files#r40075748>.
>